File: Heap/heap.cpp
#include <vector>
#include <string>
#include <iostream>
using namespace std;

class max_heap {
public:
    max_heap() {}
    max_heap(vector<int> &arr) {
        this->arr = arr;
        heapify();
    }

    void heapify() {
        vector<int> &arr = this->arr;
        for (int i = arr.size() / 2 - 1; i >= 0; --i)
            _sift(arr, i);
    }

    void _sift(vector<int> &arr, int idx) {
        int n = arr.size();
        while (1) {
            int child = idx * 2 + 1;
            if (child >= n)
                break;
            if (child + 1 < n && arr[child] < arr[child + 1])
                child++;
            if (arr[idx] < arr[child]) {
                swap(arr[idx], arr[child]);
                idx = child;
            } else {
                break;
            }
        }
    }

    void push(int value) {
        vector<int> &arr = this->arr;
        int idx = arr.size();
        arr.push_back(value);
        while (idx) {
            if (arr[(idx - 1) / 2] >= value)
                break;
            arr[idx] = arr[(idx - 1) / 2];
            idx = (idx - 1) / 2;
        }
        arr[idx] = value;
    }

    int pop() {
        vector<int> &arr = this->arr;
        int maxi = arr[0];
        arr[0] = arr.back();
        arr.pop_back();
        _sift(arr, 0);
        return maxi;
    }

    bool empty() {
        return this->arr.empty();
    }

    string print() {
        string res = "[";
        for (auto v: this->arr)
            res += to_string(v) + ", ";
        res.pop_back();
        res.pop_back();
        res += "]";
        return res;
    }

private:
    vector<int> arr;
};
int main() {
    max_heap m = max_heap();
    m.heapify();
    m.print();
}

File: Heap/heap.py

class max_heap:
    def __init__(self, arr: list[int]) -> None:
        self.arr = arr
        self.heapify()
    
    def heapify(self):
        for idx in range(len(self.arr) - 1, -1, -1):
            self._sift(idx)

    def _sift(self, idx):
        n = len(self.arr)
        while 1:
            child = idx * 2 + 1
            if child >= n:
                break
            if child + 1 < n and self.arr[child] < self.arr[child+1]:
                child += 1
            if self.arr[idx] < self.arr[child]:
                self.arr[idx], self.arr[child] = self.arr[child], self.arr[idx]
                idx = child
            else:
                break

    def push(self, value):
        idx = len(self.arr)
        self.arr.append(value)
        while idx:
            if self.arr[(idx - 1) // 2] >= value:
                break
            self.arr[idx] = self.arr[(idx - 1) // 2]
            idx = (idx - 1) // 2
        self.arr[idx] = value
    
    def top(self):
        return self.arr[0] if self.arr else -1
    
    def pop(self):
        maxi = self.arr[0]
        self.arr[0] = self.arr.pop()
        self._sift(0)
        return maxi

    def __len__(self):
        return len(self.arr)
    
    def empty(self):
        return len(self.arr) == 0